test/grove-base-fork/Curve.t.sol

```diff
@@ -0,0 +1,1337 @@
+        rateLimits.setUnlimitedRateLimitData(curveWithdrawKey);
+        vm.stopPrank();
+
+        _addLiquidity(100_000_000e6, 100_000_000e6);
+
+        uint256 virtualPrice1 = curvePool.get_virtual_price();
+
+        // Verify initial virtual price is reasonable (pool may have slight imbalance from setup)
+        assertApproxEqRel(virtualPrice1, 1.020385339110137837e18, 0.01e18);  // Within 1%
+
+        deal(address(usdcBase), address(almProxy), 100_000_000e6);
+
+        vm.prank(GROVE_EXECUTOR);
+        foreignController.setMaxSlippage(CURVE_POOL, 1);  // 1e-16%
+
+        // Calculate expected swap output dynamically
+        uint256 expectedAmountOut = curvePool.get_dy(0, 1, 100_000_000e6);
+
+        // Perform a massive swap to stress the virtual price
+        vm.prank(ALM_RELAYER);
+        uint256 amountOut = foreignController.swapCurve(CURVE_POOL, 0, 1, 100_000_000e6, 1000e6);
+
+        assertEq(amountOut, expectedAmountOut);
+
+        // Assert price rises
+        uint256 virtualPrice2 = curvePool.get_virtual_price();
+
+        assertGt(virtualPrice2, virtualPrice1);
+
+        // Add one sided liquidity to stress the virtual price
+        uint256 totalSupplyBefore = curveLp.totalSupply();
+        _addLiquidity(0, 100_000_000e6);
+        uint256 totalSupplyAfter = curveLp.totalSupply();
+
+        // Calculate expected virtual price based on the value added and LP tokens minted
+        // virtualPrice = totalPoolValue / totalSupply
+        // New expected price = (oldPrice * oldSupply + valueAdded) / (oldSupply + newLPTokens)
+        uint256[] memory rates = curvePool.stored_rates();
+        uint256 valueAdded = 100_000_000e6 * rates[1] / 1e18;  // cgUSD value in 18 decimals
+
+        uint256 expectedVirtualPrice3 = (virtualPrice2 * totalSupplyBefore + valueAdded * 1e18) / totalSupplyAfter;
+
+        uint256 virtualPrice3 = curvePool.get_virtual_price();
+
+        // Assert the calculated virtual price matches expectation (within small tolerance for rounding)
+        assertApproxEqRel(virtualPrice3, expectedVirtualPrice3, 0.001e18);  // Within 0.1%
+        assertGt(virtualPrice3, virtualPrice2);
+
+        // Remove liquidity
+        uint256[] memory minWithdrawAmounts = new uint256[](2);
+        minWithdrawAmounts[0] = 1000e6;
+        minWithdrawAmounts[1] = 1000e6;
+
+        vm.startPrank(ALM_RELAYER);
+        foreignController.removeLiquidityCurve(
+            CURVE_POOL,
+            curveLp.balanceOf(address(almProxy)),
+            minWithdrawAmounts
+        );
+        vm.stopPrank();
+
+        // Assert price rises after liquidity removal (due to fees being distributed to remaining LP holders)
+        uint256 virtualPrice4 = curvePool.get_virtual_price();
+
+        // Virtual price should continue to increase slightly due to fee distribution
+        assertGt(virtualPrice4, virtualPrice3);
+    }
+
+}
+
+// contract MainnetController3PoolSwapRateLimitTest is ForkTestBase {
```

Comment:
I assume we can get rid of this entirely?

Reply:
https://github.com/grove-labs/grove-alm-controller/pull/47/commits/e2153e36ce60fbf28bd44c40ac5ffc8e0b750bf2